In [1]:
import pandas as pd
import numpy as np

In [2]:
import matplotlib.pyplot as plt

In [3]:
df= pd.read_csv('dkmacovid_train.csv')
df_org= df.copy()


In [4]:
df.describe()

Day     State ID          Lat        Long_        Active  \
count  1380.000000  1380.000000  1380.000000  1380.000000  1.380000e+03   
mean     16.500000    25.239130    39.470717   -92.879928  2.610390e+05   
std       8.658579    14.513405     6.070494    19.632514  4.914059e+05   
min       2.000000     1.000000    21.094300  -157.498300  9.550000e+02   
25%       9.000000    12.000000    35.630100  -105.311100  2.731600e+04   
50%      16.500000    25.500000    39.583950   -88.259400  1.005915e+05   
75%      24.000000    37.000000    43.326600   -77.209800  2.592418e+05   
max      31.000000    51.000000    61.370700   -69.381900  3.283336e+06   

       Incident_Rate  Total_Test_Results  Case_Fatality_Ratio   Testing_Rate  \
count    1380.000000        1.380000e+03          1380.000000    1380.000000   
mean     7203.192905        5.271097e+06             1.631757   91763.237514   
std      2305.025102        6.991478e+06             0.656702   40858.185997   
min      1232.233261        3.739460e+05             0.439598   30524.071590   
25%      6042.134459        1.310515e+06             1.246993   67457.197525   
50%      7453.675955        2.919566e+06             1.499993   85438.613770   
75%      8621.924085        6.093790e+06             1.817013  104509.453475   
max     12811.162350        4.227902e+07             3.928767  235733.711200   

       Density Rank 2020 Census     SexRatio  
count               1380.000000  1380.000000  
mean                  27.173913    97.760870  
std                   15.378197     3.219219  
min                    1.000000    94.000000  
25%                   13.000000    95.000000  
50%                   28.500000    97.000000  
75%                   41.000000    99.000000  
max                   52.000000   109.000000

                              Table 1 Original database 

Outliers are found using the Z-score method and the Interquantile method(1). The method that finds the most outliers accurately is then choosen. Numerical features including ”Day”, and ”State ID” are checked to find if any outliers exist. Non-numerical features such as ”State” are not used in searching for outliers.

# Finding Outliers Using Z-score for Each Feature:

Outliers are found using the Z-score method for each feature that has numeric values using the method defined below.

In [5]:
outliers=[]

In [6]:
def detect_outliers(data):
    
    threshold= 3
    mean= np.mean(data)
    std= np.std(data)
    
    for i in data:
        z_score= (i-mean)/std
        
        if np.abs(z_score)> threshold:
            
            outliers.append(i)
            
            return outliers
    

In [7]:
outliers_pt= detect_outliers(df['Day'])
print(outliers_pt)

None


In [8]:
outliers_pt= detect_outliers(df['State ID'])
print(outliers_pt)

None


In [9]:
outliers_pt1= detect_outliers(df['Lat'])
print(outliers_pt1)

[61.3707]


In [10]:
outliers_pt2= detect_outliers(df['Long_'])
print(outliers_pt2)

[61.3707, -152.4044]


In [11]:
outliers_pt3= detect_outliers(df['Active'])
print(outliers_pt3)

[61.3707, -152.4044, 2362015]


In [12]:
outliers_pt4= detect_outliers(df['Incident_Rate'])
print(outliers_pt4)

None


In [13]:
outliers_pt5= detect_outliers(df['Total_Test_Results'])
print(outliers_pt5)

[61.3707, -152.4044, 2362015, 33391442]


In [14]:
outliers_pt5= detect_outliers(df['Case_Fatality_Ratio'])
print(outliers_pt5)

[61.3707, -152.4044, 2362015, 33391442, 3.928767415]


In [15]:
outliers_pt6= detect_outliers(df['Testing_Rate'])
print(outliers_pt6)

[61.3707, -152.4044, 2362015, 33391442, 3.928767415, 216923.3151]


In [16]:
outliers_pt6= detect_outliers(df['Density Rank 2020 Census'])
print(outliers_pt6)

None


In [17]:
outliers_pt7= detect_outliers(df['SexRatio'])
print(outliers_pt7)

[61.3707, -152.4044, 2362015, 33391442, 3.928767415, 216923.3151, 109]


# Finding Outliers using Interquantile Range:

Outliers are found using the Interquantile Range method for each feature that has numeric values using the method defined below.

In [18]:
 quantile1, quantile3= np.percentile(df['State ID'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)


-25.5 74.5


In [19]:
 quantile1, quantile3= np.percentile(df['Day'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)


-13.5 46.5


'State ID' and 'Day' does not have any outliers as its numerical values fall under its respective upper and lower bound ranges.

In [20]:
 
     for x in ['Lat']:

    quantile1, quantile3= np.percentile(df['Lat'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1- (1.5 * iqr_valu)
uboundval= quantile3+ (1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan


24.08535 54.87135


Outliers are present in 'Lat' after the upper bound value 54.87135 as the max value is 61.37 in specified data above. 

In [21]:
     for x in ['Long_']:

    quantile1, quantile3= np.percentile(df['Long_'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

-147.46304999999998 -35.05785000000001


Outliers are present in 'Long_' as the minimum value for this feature is -157.4983 which is less than the lower bound value -147.4630  as specified above.

In [22]:
      for x in ['Active']:

    quantile1, quantile3= np.percentile(df['Active'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan



-320572.625 607130.375


Outliers are present in 'Lat' after the upper bound value 607130.375 as the max value is 3.283336e+06 as specified above.

In [23]:
       for x in ['Incident_Rate']:

    quantile1, quantile3= np.percentile(df['Incident_Rate'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

2172.450020124999 12491.608523125002


Outliers are present in 'Incident_Rate' below the lower bound value and also above the upper bounds 2172.450020124999, 12491.608523125002 respectively as the min and max values are 1232.233261 and 12811.162350 respectively.

In [24]:
        for x in ['Total_Test_Results']:

    quantile1, quantile3= np.percentile(df['Total_Test_Results'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

-5864398.5 13268703.5


Outliers are present in 'Total_Test_Results' after the upper bound value 607130.375 as the max value is 4.227902e+07 in the numerics table specified above.

In [25]:
        for x in ['Case_Fatality_Ratio']:

    quantile1, quantile3= np.percentile(df['Case_Fatality_Ratio'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

0.3919640299999998 2.67204191


Outliers are present in 'Total_Test_Results' after the upper bound value 2.67204191 as the max value is 3.928767 in the numerics table specified above.

In [26]:
        for x in ['Testing_Rate']:

    quantile1, quantile3= np.percentile(df['Testing_Rate'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

11878.813599999987 160087.83740000002


Outliers are present in 'Testing_Rate' below the lower bound value and also above the upper bounds 11878.813599999987, 160087.83740000002 respectively as the min and max values are 30524.071590 and 235733.711200 respectively.

In [27]:
        for x in ['Density Rank 2020 Census']:

    quantile1, quantile3= np.percentile(df['Density Rank 2020 Census'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

-29.0 83.0


'Density Rank 2020 Census' does not have any outliers as it's min and max(1.000000,52.000000) values fall within their respective upper and lower bound ranges(-29.0, 83.0) in the numerics table specified above.

In [28]:
        for x in ['SexRatio']:

    quantile1, quantile3= np.percentile(df['SexRatio'],[25,75])

iqr_valu=quantile3- quantile1

lboundval= quantile1-(1.5 * iqr_valu)
uboundval= quantile3+(1.5 * iqr_valu)

print(lboundval,uboundval)

df.loc[df[x]< lboundval,x]= np.nan
df.loc[df[x]> uboundval,x]= np.nan

89.0 105.0


Outliers are present in 'SexRatio' after the upper bound value 105.0 as the max value is 109.000000 in the numerics table specified above.

In [29]:
df=df.dropna(axis=0)
df.describe()

Day     State ID          Lat        Long_         Active  \
count  1003.000000  1003.000000  1003.000000  1003.000000    1003.000000   
mean     16.514457    26.936191    39.317259   -92.393330  135733.039880   
std       8.604445    14.160539     4.092189    13.825014  134797.128608   
min       2.000000     1.000000    31.169500  -122.070900    1098.000000   
25%       9.000000    17.000000    35.630100  -105.311100   27737.500000   
50%      17.000000    27.000000    39.063900   -91.867800   76981.000000   
75%      24.000000    38.000000    43.326600   -80.945000  209076.500000   
max      31.000000    51.000000    46.921900   -69.381900  606125.000000   

       Incident_Rate  Total_Test_Results  Case_Fatality_Ratio   Testing_Rate  \
count    1003.000000        1.003000e+03          1003.000000    1003.000000   
mean     7387.058370        3.515933e+06             1.506624   78252.900036   
std      1975.065568        2.531671e+06             0.476659   25477.512412   
min      2179.566914        3.739460e+05             0.454009   30524.071590   
25%      6108.339399        1.303540e+06             1.223279   63815.618495   
50%      7722.730108        2.652670e+06             1.442444   78804.952950   
75%      8739.907284        5.464123e+06             1.649429   96924.767205   
max     12236.353220        9.366357e+06             2.667652  155928.807600   

       Density Rank 2020 Census     SexRatio  
count               1003.000000  1003.000000  
mean                  30.328016    97.710867  
std                   13.546368     2.576601  
min                    1.000000    94.000000  
25%                   20.000000    96.000000  
50%                   31.000000    98.000000  
75%                   42.000000    99.000000  
max                   51.000000   104.000000

Table 2 Corrected database after removing outliers

 The outliers are assigned a null value and are then dropped(2):

In [30]:
df.isnull().sum()

Day                                0
State ID                           0
State                              0
Lat                                0
Long_                              0
Active                             0
Incident_Rate                      0
Total_Test_Results                 0
Case_Fatality_Ratio                0
Testing_Rate                       0
Resident Population 2020 Census    0
Population Density 2020 Census     0
Density Rank 2020 Census           0
SexRatio                           0
Confirmed                          0
Deaths                             0
Recovered                          0
dtype: int64

# Z-Score Normalization:

In [31]:
def zscore(a): 
    
    xx=a
    a_std = (xx- np.mean(xx)) / (np.std(xx))
    return a_std

The Z-Score normilization is then applied after the outliers are removed.

In [32]:
df['Day']= zscore(df['Day'])
df['State ID']= zscore(df['State ID'])
df['Lat']= zscore(df['Lat'])
df['Long_']= zscore(df['Long_'])
df['Active']= zscore(df['Active'])
df['Incident_Rate']= zscore(df['Incident_Rate'])
df['Total_Test_Results']= zscore(df['Total_Test_Results'])
df['Case_Fatality_Ratio']= zscore(df['Case_Fatality_Ratio'])
df['Testing_Rate']= zscore(df['Testing_Rate'])
df['Density Rank 2020 Census']= zscore(df['Density Rank 2020 Census'])
df['SexRatio']= zscore(df['SexRatio'])

In [33]:
df.describe()

Day      State ID           Lat         Long_        Active  \
count  1003.000000  1.003000e+03  1.003000e+03  1.003000e+03  1.003000e+03   
mean      0.000000 -4.958922e-17  2.320067e-16 -2.125252e-17 -8.855219e-17   
std       1.000499  1.000499e+00  1.000499e+00  1.000499e+00  1.000499e+00   
min      -1.687697 -1.832496e+00 -1.992045e+00 -2.147728e+00 -9.992958e-01   
25%      -0.873758 -7.020318e-01 -9.014731e-01 -9.348427e-01 -8.015706e-01   
50%       0.056458  4.508332e-03 -6.194377e-02  3.803197e-02 -4.360727e-01   
75%       0.870396  7.817024e-01  9.802433e-01  8.285013e-01  5.443740e-01   
max       1.684335  1.700205e+00  1.859258e+00  1.665308e+00  3.491370e+00   

       Incident_Rate  Total_Test_Results  Case_Fatality_Ratio  Testing_Rate  \
count   1.003000e+03        1.003000e+03         1.003000e+03  1.003000e+03   
mean   -1.983569e-16        7.792592e-17        -1.328283e-16 -2.833670e-17   
std     1.000499e+00        1.000499e+00         1.000499e+00  1.000499e+00   
min    -2.637932e+00       -1.241691e+00        -2.209422e+00 -1.874305e+00   
25%    -6.477541e-01       -8.743223e-01        -5.947375e-01 -5.669503e-01   
50%     1.700395e-01       -3.411554e-01        -1.347131e-01  2.167905e-02   
75%     6.853058e-01        7.699113e-01         2.997445e-01  7.332420e-01   
max     2.456483e+00        2.312047e+00         2.436978e+00  3.050324e+00   

       Density Rank 2020 Census      SexRatio  
count              1.003000e+03  1.003000e+03  
mean              -3.387121e-17 -2.503813e-16  
std                1.000499e+00  1.000499e+00  
min               -2.166090e+00 -1.440936e+00  
25%               -7.627999e-01 -6.643329e-01  
50%                4.963096e-02  1.122707e-01  
75%                8.620619e-01  5.005725e-01  
max                1.526778e+00  2.442081e+00

Table 3 Corrected database after removing outliers and applying zscore normalization 

In [34]:
df.to_csv('df-filteredfinalN.csv',index=False)

# Conclusion:

In conclusion, the interquantile range is chosen over the z-score for outlier detection as it was more accurate in detecting outliers by accounting for the upper and lower bounds of the data by considering the 25th and 75th percentile of the features in the data.The Interquantile method was able to find even the smallest range outliers such as with the feature 'Incident Rate' which was not found with the Z-score method. The outliers are assigned a null value and are then dropped after finding them through the interquantile method. The filtered data is normalized using Min-Max normalization and then downloaded as a new csv file for the following questions.

# References:

(1) https://www.youtube.com/watch?v=rzR_cKnkD18

(2) https://www.askpython.com/python/examples/detection-removal-outliers-in-python